# Workflow: 

- Step 1.  For 'Position_clean' column, find all rows that contain substring combo ('intern' OR 'co-op) AND ('data science' OR 'data scientist' OR 'ML' or 'machine learning') *case insensitive*
    - replace all with 'Data Scientist Intern' 
- Step 2.  find all rows that contain substring 'data science'*case insensitive*
    - replace case by case with 'xx Data Scientist' 
- Step 3. find all rows that contain substring 'analytics' *case insensitive*
    - replace case by case with 'xx Data Scientist' 
- Step 4. find all rows that contain substring 'ML' or 'machine learning' (assume: exclude 'AI' roles) *case-insensitive*
    - replace case by case with 'xx Data Scientist' 
 
- Step 5. Other special cases (from examining df.Position. unique()): replace case by case with 'xx Data Scientist' 
  
- Step 6: find all rows that do NOT contain substring *case-insensitive* 'Data Scientist'(e.g. 'software engineer', 'AI research scientist', 'data engineer', 'software developer', 'database developer', 'data analyst') -> replace with np.nan (for easier removal of these rows later) 


[see here for str.replace](https://stackoverflow.com/questions/39768547/replace-whole-string-if-it-contains-substring-in-pandas)

**note: order matters! e.g. '*Senior* *director*, data scientist' - need to change to LDS, before SR**

- Step 7A. replace any substrings 'lead' or 'director' or 'manager' or 'principal' -> 'Lead Data Scientist'
- Step 7B. replace any substrings 'sr' or 'senior' -> 'Senior Data Scientist'
- Step 7C. replace any substrings jr' or 'junior' -> 'Junior Data Scientist'
- Step 7D. replace substrings NOT 'intern' NOT 'junior' NOT 'senior' NOT 'lead' NOT NaN  -> 'Data Scientist' 


AFTER merging with Sasha's cleaned columns, do following: 
- drop rows that have NaN in the 'Position_clean' column [see how here](https://stackoverflow.com/questions/13413590/how-to-drop-rows-of-pandas-dataframe-whose-value-in-a-certain-column-is-nan)
- check and drop duplicates
- drop other rows with certain NaN fields

In [34]:
import pandas as pd
import numpy as np

In [2]:
INTERN = 'Data Scientist Intern'
JR = 'Junior Data Scientist'
SR = 'Senior Data Scientist'
DS = 'Data Scientist'
LDS = 'Lead Data Scientist'
rm = np.nan

In [3]:
df_ca = pd.read_csv('jack_canada.csv', index_col = 0)

In [4]:
df_ca.shape

(1258, 12)

In [5]:
df_ca.head()

Company  \
0                             AlayaCare   
1  Special Risk Insurance Managers Ltd.   
2                                Syniti   
3                  Points International   
4                       Spin Master Ltd   

                                         Description  \
0  Do you believe that recent advancements in AI ...   
1  Be a part of a dynamic and growing team at Spe...   
2  Syniti was founded to solve business’ complex ...   
3  Company Description\r\r\r\r\r\r\r\r\r\r\r\r\r\...   
4  Are you looking to further your career with an...   

   Glassdoor Salary Estimate                                 Industry  \
0                        NaN  Enterprise Software & Network Solutions   
1                        NaN                                      NaN   
2                        NaN                              IT Services   
3                        NaN                                 Internet   
4                        NaN          Consumer Products Manufacturing   

   Location                      Position  Rating                     Revenue  \
0  Montreal                Data Scientist     4.4    Unknown / Non-Applicable   
1   Langley                  Data Analyst     NaN                         NaN   
2   Moncton                  Data Analyst     3.7  $100 to $500 million (USD)   
3   Toronto                Data Scientist     3.6  $100 to $500 million (USD)   
4   Toronto  Senior Manager, Data Science     3.0      $2 to $5 billion (USD)   

                   Sector                    Size               Type  \
0  Information Technology    201 to 500 Employees  Company - Private   
1                     NaN                     NaN                NaN   
2  Information Technology   501 to 1000 Employees  Company - Private   
3  Information Technology    201 to 500 Employees   Company - Public   
4           Manufacturing  1001 to 5000 Employees   Company - Public   

   Year Founded  
0        2014.0  
1           NaN  
2        1996.0  
3        2001.0  
4        1994.0

In [6]:
# duplicate
df_ca['Position_clean'] = df_ca.loc[:, 'Position']

In [7]:
df_ca['Position'].unique()

array(['Data Scientist', 'Data Analyst', 'Senior Manager, Data Science',
       'Automotive Software Data Analyst / Tester - Remote',
       'Junior Data Analyst',
       'Optimization Operational Research Scientist',
       'Data Scientist Intern', 'Junior Scientist',
       'Senior Database Developer', 'Master Data Analyst / SAP (Junior)',
       'Data Analyst - Junior', 'Data Analyst (Remote)',
       'Remote PT Data Science Mentor',
       'Remote Cloud Database developer (Python)',
       'Reporting Consultant & Data Analyst / Cognos Expert',
       'Flight Data Analyst', 'Junior Software Developer',
       'Product Data Analyst', 'Data Analyst II',
       'Entry Level Data Analyst (Bilingual French & English)',
       'Jr. Data Analyst', 'Digital Data Analyst',
       'Principal Data Scientist', 'Data Analyst, Security - 297742',
       'Tech - Data Analyst', 'Data Analyst, Product',
       'Data Analyst Intern, Global Investments IT',
       'Data Analyst - Level 1 x2 (Contract)

In [8]:
df_ca[df_ca['Position'] == 'Senior Statistical Programmer/Statistical Scientist']

Company                                        Description  \
692   IQVIA  IQVIA™ is the leading human data science compa...   

     Glassdoor Salary Estimate                   Industry  Location  \
692                        NaN  Biotech & Pharmaceuticals  Kirkland   

                                              Position  Rating  \
692  Senior Statistical Programmer/Statistical Scie...     3.7   

                    Revenue                     Sector              Size  \
692  $2 to $5 billion (USD)  Biotech & Pharmaceuticals  10000+ Employees   

                 Type  Year Founded  \
692  Company - Public        2017.0   

                                        Position_clean  
692  Senior Statistical Programmer/Statistical Scie...

In [9]:
## step 1
s = ['intern', 'co-op', 'co op', 'placem']
cond_or1 = '|'.join(s)
t = ["data scien", "machine learn"]
cond_or2 = '|'.join(t)
bool_intern = (df_ca['Position'].str.contains(cond_or1, case = False))& \
               ( (df_ca['Position'].str.contains(cond_or2, case = False)) |\
                (df_ca['Position'].str.contains('ML', case = True))
               ) 
df_ca[bool_intern] # display rows

Company  \
10                                   Guavus   
286                               Reesby IT   
370                                  Guavus   
1056                                    RBC   
1159                               QuadReal   
1174                                 Guavus   
1181  Southern Graphics Systems, Canada Co.   
1208                                  sgsco   
1210                American Iron and Metal   

                                            Description  \
10    Guavus, a Thales company, is seeking a Data Sc...   
286   About the Data Scientist Internship\r\r\r\r\r\...   
370   Guavus, a Thales company, is seeking a Data Sc...   
1056  What is the opportunity?\r\r\r\r\r\r\r\r\n\r\r...   
1159  About QuadReal\r\r\r\r\r\nHeadquartered in Van...   
1174  Guavus, a Thales company, is seeking a Data Sc...   
1181  Data Scientist Intern– sgsco labs\r\r\r\r\n\r\...   
1208  Data Scientist Intern– sgsco labs\r\r\r\n\r\r\...   
1210  Company Description\r\r\r\n\r\r\r\nES-TU FORGÉ...   

      Glassdoor Salary Estimate                                 Industry  \
10                          NaN  Enterprise Software & Network Solutions   
286                         NaN                                      NaN   
370                         NaN  Enterprise Software & Network Solutions   
1056                        NaN                    Banks & Credit Unions   
1159                        NaN                              Real Estate   
1174                        NaN  Enterprise Software & Network Solutions   
1181                        NaN                  Advertising & Marketing   
1208                        NaN                  Advertising & Marketing   
1210                        NaN            Metal & Mineral Manufacturing   

              Location                                           Position  \
10            Montreal                              Data Scientist Intern   
286   British Columbia  Machine Learning Artificial Intelligence Inter...   
370           Montreal                              Data Scientist Intern   
1056           Toronto  Data Scientist - Winter 2021 Internship (8 Mon...   
1159           Toronto                             Intern, Data Scientist   
1174          Montreal                              Data Scientist Intern   
1181           Toronto                              Intern - Data Science   
1208           Toronto                              Intern - Data Science   
1210      Montréal-Est                          Internship Data Scientist   

      Rating                   Revenue                  Sector  \
10       2.8  $25 to $50 million (USD)  Information Technology   
286      NaN                       NaN                     NaN   
370      2.8  $25 to $50 million (USD)  Information Technology   
1056     4.0        $10+ billion (USD)                 Finance   
1159     4.4  Unknown / Non-Applicable             Real Estate   
1174     2.8  $25 to $50 million (USD)  Information Technology   
1181     3.0  Unknown / Non-Applicable       Business Services   
1208     2.5  Unknown / Non-Applicable       Business Services   
1210     2.6    $1 to $2 billion (USD)           Manufacturing   

                        Size                            Type  Year Founded  \
10      201 to 500 Employees  Subsidiary or Business Segment        2006.0   
286                      NaN                             NaN           NaN   
370     201 to 500 Employees  Subsidiary or Business Segment        2006.0   
1056        10000+ Employees                Company - Public        1869.0   
1159  1001 to 5000 Employees               Company - Private        2016.0   
1174    201 to 500 Employees  Subsidiary or Business Segment        2006.0   
1181  1001 to 5000 Employees  Subsidiary or Business Segment        1946.0   
1208  1001 to 5000 Employees               Company - Private        1946.0   
1210  1001 to 5000 Employees               Company - Private        1936.0   



In [10]:
df_ca.loc[bool_intern, 'Position_clean'] = INTERN
# 1c. check by comparing 2 columns
df_ca[bool_intern][['Position', 'Position_clean']]

Position         Position_clean
10                                Data Scientist Intern  Data Scientist Intern
286   Machine Learning Artificial Intelligence Inter...  Data Scientist Intern
370                               Data Scientist Intern  Data Scientist Intern
1056  Data Scientist - Winter 2021 Internship (8 Mon...  Data Scientist Intern
1159                             Intern, Data Scientist  Data Scientist Intern
1174                              Data Scientist Intern  Data Scientist Intern
1181                              Intern - Data Science  Data Scientist Intern
1208                              Intern - Data Science  Data Scientist Intern
1210                          Internship Data Scientist  Data Scientist Intern

In [11]:
## Step 2
bool_2 = df_ca['Position'].str.contains('data science', case = False)
df_ca[bool_2] # display

Company  \
4                           Spin Master Ltd   
75                          Lighthouse Labs   
204                Loblaw Companies Limited   
210                                  Loblaw   
286                               Reesby IT   
363                         Spin Master Ltd   
393                         Spin Master Ltd   
395                             Bell Canada   
459                                    Citi   
474                                Citibank   
544                                  Loblaw   
738                                   Aviva   
857                Loblaw Companies Limited   
868                         Lighthouse Labs   
878                                  Intact   
923                                     RBC   
968                Loblaw Companies Limited   
969                                  Loblaw   
970                Loblaw Companies Limited   
984                                  Loblaw   
1181  Southern Graphics Systems, Canada Co.   
1208                                  sgsco   

                                            Description  \
4     Are you looking to further your career with an...   
75    *DUE TO COVID WE ARE CURRENTLY FULLY REMOTE AN...   
204   Lieu:\r\r\r\r\r\r\r\r\n1 Presidents Choice Cir...   
210   Location:\r\r\r\r\r\r\r\n\r\r\r\r\r\r\r\n1 Pre...   
286   About the Data Scientist Internship\r\r\r\r\r\...   
363   Are you looking to further your career with an...   
393   Are you looking to further your career with an...   
395   Req Id: 272559\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r...   
459   The Applications Development Senior Manager ac...   
474   The Applications Development Senior Manager ac...   
544   Location:\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\...   
738   ** FUTURE OPPORTUNITY **\r\r\r\r\r\r\r\r\r\r\r...   
857   Lieu:\r\r\r\r\r\r\r\r\r\r\r\r\r\r\r\n1 Preside...   
868   *DUE TO COVID WE ARE CURRENTLY FULLY REMOTE AN...   
878   How do you imagine life at an insurance compan...   
923   What is the opportunity?\r\r\r\r\r\r\r\r\r\r\r...   
968   Lieu:\r\r\r\r\r\r\r\r\r\r\r\n1 Presidents Choi...   
969   Location:\r\r\r\r\r\r\r\r\r\r\r\n\r\r\r\r\r\r\...   
970   Lieu:\r\r\r\r\r\r\r\r\r\r\r\n1 Presidents Choi...   
984   Location:\r\r\r\r\r\r\r\r\r\r\r\n\r\r\r\r\r\r\...   
1181  Data Scientist Intern– sgsco labs\r\r\r\r\n\r\...   
1208  Data Scientist Intern– sgsco labs\r\r\r\n\r\r\...   

      Glassdoor Salary Estimate                               Industry  \
4                           NaN        Consumer Products Manufacturing   
75                          NaN            Education Training Services   
204                         NaN                        Vehicle Dealers   
210                         NaN                        Vehicle Dealers   
286                         NaN                                    NaN   
363                         NaN        Consumer Products Manufacturing   
393                         NaN        Consumer Products Manufacturing   
395                         NaN  Cable, Internet & Telephone Providers   
459                         NaN  Investment Banking & Asset Management   
474                         NaN                                Lending   
544                         NaN                        Vehicle Dealers   
738                         NaN                     Insurance Carriers   
857                         NaN                        Vehicle Dealers   
868                         NaN            Education Training Services   
878                         NaN                     Insurance Carriers   
923                         NaN                  Banks & Credit Unions   
968                         NaN                        Vehicle Dealers   
969                         NaN                        Vehicle Dealers   
970                         NaN                        Vehicle Dealers   
984                         NaN                        Vehicle Dealers   
1181                        NaN           

In [12]:
# special
df_ca.loc[df_ca['Position'] == 'Remote PT Data Science Mentor', 'Position_clean' ] = rm

In [13]:
# lead ds
bool_lds = (df_ca['Position_clean'].str.contains('Manage|Direct|Lead', case = False)) & \
           (df_ca['Position_clean'].str.contains('data science', case = False))
df_ca.loc[bool_lds, 'Position_clean'] = LDS
# sr ds
bool_sr = (df_ca['Position_clean'].str.contains('Senior|Sr', case = False)) &\
            (df_ca['Position_clean'].str.contains('data science', case = False))
df_ca.loc[bool_sr, 'Position_clean'] = SR
# ds
bool_ds = (df_ca['Position_clean'].str.contains('Specialist|Analyst', case = False)) &\
            (df_ca['Position_clean'].str.contains('data science', case = False))
df_ca.loc[bool_ds, 'Position_clean'] = DS

In [14]:
# check 
df_ca[bool_2][['Position', 'Position_clean']]

Position         Position_clean
4                          Senior Manager, Data Science    Lead Data Scientist
75                        Remote PT Data Science Mentor                    NaN
204                              Director, Data Science    Lead Data Scientist
210                   Director, Data Science - AIML Coe    Lead Data Scientist
286   Machine Learning Artificial Intelligence Inter...  Data Scientist Intern
363                        Senior Manager, Data Science    Lead Data Scientist
393                        Senior Manager, Data Science    Lead Data Scientist
395                        Senior Manager, Data Science    Lead Data Scientist
459                         Data Science Senior Manager    Lead Data Scientist
474                         Data Science Senior Manager    Lead Data Scientist
544                   Director, Data Science - AIML Coe    Lead Data Scientist
738   Senior Actuarial Analyst, Commercial Insurance...  Senior Data Scientist
857                              Director, Data Science    Lead Data Scientist
868                       Remote PT Data Science Mentor                    NaN
878                 Data Science Specialist, Intact Lab         Data Scientist
923                              Director, Data Science    Lead Data Scientist
968         Data Science Analyst,Customer Analytics/CRM         Data Scientist
969   Data Science Analyst,Customer Analytics/CRM - ...         Data Scientist
970        Data Science Analyst, Customer Analytics/CRM         Data Scientist
984   Data Science Analyst, Customer Analytics/CRM -...         Data Scientist
1181                              Intern - Data Science  Data Scientist Intern
1208                              Intern - Data Science  Data Scientist Intern

In [15]:
## Step 3
# display all 'analytics' roles
bool_3 = df_ca['Position'].str.contains('analytics', case = False)
df_ca[bool_3][['Position', 'Position_clean']]# display

Position  \
269   Assistant Manager (Senior Data Analyst), Decis...   
290                              Manager Data Analytics   
420           Data Scientist, Marketing & Web Analytics   
435        Lead Data Scientist (Supply Chain Analytics)   
456   Specialist System Design, Data Scientist Machi...   
506                     Data Scientist, Video Analytics   
509                  Data Scientist, Advanced Analytics   
513                  Senior Data Scientist BI Analytics   
519                    Senior Data Engineer - Analytics   
533                Senior Data Scientist – BI Analytics   
562            Senior Software Engineer, Data Analytics   
606                              Manager Data Analytics   
647             Senior Software Engineer-Data Analytics   
697   Spécialiste, System Design, Data Scientist Mac...   
720             Analytics Engineering Senior Specialist   
722                   Senior Manager Advanced Analytics   
728                    Analytics DevOps Senior Engineer   
734                   Data & Analytics Strategy Manager   
743      Senior Product Manager, Retail Analytics (EST)   
751   Assistant Manager (Senior Data Analyst), Decis...   
797                Product Director, Data and Analytics   
800           Senior Actuarial Analyst - Risk Analytics   
842     Data and Analytics Quantitative Analyst 3157309   
845                    Data Scientist, People Analytics   
887   Intermediate/Senior Backend Developer, Analyti...   
951              Data & Analytics - Senior Data Analyst   
968         Data Science Analyst,Customer Analytics/CRM   
969   Data Science Analyst,Customer Analytics/CRM - ...   
970        Data Science Analyst, Customer Analytics/CRM   
984   Data Science Analyst, Customer Analytics/CRM -...   
1144            Data and Analytics Quantitative Analyst   

                                         Position_clean  
269   Assistant Manager (Senior Data Analyst), Decis...  
290                              Manager Data Analytics  
420           Data Scientist, Marketing & Web Analytics  
435        Lead Data Scientist (Supply Chain Analytics)  
456   Specialist System Design, Data Scientist Machi...  
506                     Data Scientist, Video Analytics  
509                  Data Scientist, Advanced Analytics  
513                  Senior Data Scientist BI Analytics  
519                    Senior Data Engineer - Analytics  
533                Senior Data Scientist – BI Analytics  
562            Senior Software Engineer, Data Analytics  
606                              Manager Data Analytics  
647             Senior Software Engineer-Data Analytics  
697   Spécialiste, System Design, Data Scientist Mac...  
720             Analytics Engineering Senior Specialist  
722                   Senior Manager Advanced Analytics  
728                    Analytics DevOps Senior Engineer  
734                   Data & Analytics Strategy Manager  
743      Senior Product Manager, Retail Analytics (EST)  
751   Assistant Manager (Senior Data Analyst), Decis...  
797                Product Director, Data and Analytics  
800           Senior Actuarial Analyst - Risk Analytics  
842     Data and Analytics Quantitative Analyst 3157309  
845                    Data Scientist, People Analytics  
887   Intermediate/Senior Backend Developer, Analyti...  
951              Data & Analytics - Senior Data Analyst  
968                                      Data Scientist  
969                                      Data Scientist  
970                                      Data Scientist  
984                                      Data Scientist  
1144            Data and Analytics Quantitative Analyst

In [16]:
# replace 'engineer/PM' & 'analytics' roles with NaN
bool_eng = df_ca['Position_clean'].str.contains('engineer|product manager|backend', case = False) & \
           df_ca['Position_clean'].str.contains('analytics', case = False)
df_ca.loc[bool_eng , 'Position_clean'] = rm

In [17]:
df_ca[bool_eng][['Position', 'Position_clean']]

Position Position_clean
519                   Senior Data Engineer - Analytics            NaN
562           Senior Software Engineer, Data Analytics            NaN
647            Senior Software Engineer-Data Analytics            NaN
720            Analytics Engineering Senior Specialist            NaN
728                   Analytics DevOps Senior Engineer            NaN
743     Senior Product Manager, Retail Analytics (EST)            NaN
887  Intermediate/Senior Backend Developer, Analyti...            NaN

In [18]:
## replace 
# LDS
bool_lds = (df_ca['Position_clean'].str.contains('Manage|Direct|Lead|VP|head', case = False)) &\
           (df_ca['Position_clean'].str.contains('Analytics|Data scientist', case = False))
df_ca.loc[bool_lds, 'Position_clean'] = LDS
# SR
bool_sr = (df_ca['Position_clean'].str.contains('senior|sr', case = False)) &\
           (df_ca['Position_clean'].str.contains('Analytics|Data scientist', case = False))
df_ca.loc[bool_sr, 'Position_clean'] = SR
# DS
bool_ds = (df_ca['Position_clean'].str.contains('analyst|specialist', case = False)) &\
           (df_ca['Position_clean'].str.contains('Analytics|Data scientist', case = False))
df_ca.loc[bool_ds, 'Position_clean'] = DS
# check
df_ca[bool_3][['Position', 'Position_clean']]

Position  \
269   Assistant Manager (Senior Data Analyst), Decis...   
290                              Manager Data Analytics   
420           Data Scientist, Marketing & Web Analytics   
435        Lead Data Scientist (Supply Chain Analytics)   
456   Specialist System Design, Data Scientist Machi...   
506                     Data Scientist, Video Analytics   
509                  Data Scientist, Advanced Analytics   
513                  Senior Data Scientist BI Analytics   
519                    Senior Data Engineer - Analytics   
533                Senior Data Scientist – BI Analytics   
562            Senior Software Engineer, Data Analytics   
606                              Manager Data Analytics   
647             Senior Software Engineer-Data Analytics   
697   Spécialiste, System Design, Data Scientist Mac...   
720             Analytics Engineering Senior Specialist   
722                   Senior Manager Advanced Analytics   
728                    Analytics DevOps Senior Engineer   
734                   Data & Analytics Strategy Manager   
743      Senior Product Manager, Retail Analytics (EST)   
751   Assistant Manager (Senior Data Analyst), Decis...   
797                Product Director, Data and Analytics   
800           Senior Actuarial Analyst - Risk Analytics   
842     Data and Analytics Quantitative Analyst 3157309   
845                    Data Scientist, People Analytics   
887   Intermediate/Senior Backend Developer, Analyti...   
951              Data & Analytics - Senior Data Analyst   
968         Data Science Analyst,Customer Analytics/CRM   
969   Data Science Analyst,Customer Analytics/CRM - ...   
970        Data Science Analyst, Customer Analytics/CRM   
984   Data Science Analyst, Customer Analytics/CRM -...   
1144            Data and Analytics Quantitative Analyst   

                                         Position_clean  
269                                 Lead Data Scientist  
290                                 Lead Data Scientist  
420           Data Scientist, Marketing & Web Analytics  
435                                 Lead Data Scientist  
456                                      Data Scientist  
506                     Data Scientist, Video Analytics  
509                  Data Scientist, Advanced Analytics  
513                               Senior Data Scientist  
519                                                 NaN  
533                               Senior Data Scientist  
562                                                 NaN  
606                                 Lead Data Scientist  
647                                                 NaN  
697   Spécialiste, System Design, Data Scientist Mac...  
720                                                 NaN  
722                                 Lead Data Scientist  
728                                                 NaN  
734                                 Lead Data Scientist  
743                                                 NaN  
751                                 Lead Data Scientist  
797                                 Lead Data Scientist  
800                               Senior Data Scientist  
842                                      Data Scientist  
845                    Data Scientist, People Analytics  
887                                                 NaN  
951                               Senior Data Scientist  
968                                      Data Scientist  
969                                      Data Scientist  
970                                      Data Scientist  
984                                      Data Scientist  
1144                                     Data Scientist

In [19]:
## Step 4
bool_ml = df_ca['Position'].str.contains('machine learn', case = False, na = False) &\
         ( df_ca['Position'].str.contains('data', case = False, na = False) |\
           df_ca['Position'].str.contains('ML', case = True, na = False) )
df_ca[bool_ml][['Position', 'Position_clean']]

Position  \
286   Machine Learning Artificial Intelligence Inter...   
450             Senior Data Scientist, Machine Learning   
456   Specialist System Design, Data Scientist Machi...   
547           Data Scientist/ Machine learning Engineer   
548           Data Scientist/ Machine learning Engineer   
549           Data Scientist/ Machine learning Engineer   
697   Spécialiste, System Design, Data Scientist Mac...   
718            Data Scientist/Machine Learning Engineer   
919                   Data Scientist – Machine Learning   
1025           Machine Learning Engineer/Data Scientist   
1045  Machine Learning Engineer, Content Intelligenc...   

                                         Position_clean  
286                               Data Scientist Intern  
450                               Senior Data Scientist  
456                                      Data Scientist  
547           Data Scientist/ Machine learning Engineer  
548           Data Scientist/ Machine learning Engineer  
549           Data Scientist/ Machine learning Engineer  
697   Spécialiste, System Design, Data Scientist Mac...  
718            Data Scientist/Machine Learning Engineer  
919                   Data Scientist – Machine Learning  
1025           Machine Learning Engineer/Data Scientist  
1045  Machine Learning Engineer, Content Intelligenc...

In [20]:
# replace special case: DS/ML engineer
SPECIAL = ['Data Scientist/', 'Engineer/Data Scientist']
cond = '|'.join(SPECIAL)
df_ca.loc[df_ca['Position'].str.contains(cond, case = False, na = False), 'Position_clean'] = DS

In [21]:
# replace 'developer/engineering' & 'ML/machine learning' roles w NaN
bool_eng2 = df_ca['Position_clean'].str.contains('engineer|develop', case = False) & \
           df_ca['Position_clean'].str.contains('ML|machine learn|AI|artificial intelli', case = False) 
df_ca[bool_eng2][['Position', 'Position_clean']]
df_ca.loc[ bool_eng2 , 'Position_clean'] = rm

In [22]:
# check 
df_ca[bool_ml][['Position', 'Position_clean']]

Position  \
286   Machine Learning Artificial Intelligence Inter...   
450             Senior Data Scientist, Machine Learning   
456   Specialist System Design, Data Scientist Machi...   
547           Data Scientist/ Machine learning Engineer   
548           Data Scientist/ Machine learning Engineer   
549           Data Scientist/ Machine learning Engineer   
697   Spécialiste, System Design, Data Scientist Mac...   
718            Data Scientist/Machine Learning Engineer   
919                   Data Scientist – Machine Learning   
1025           Machine Learning Engineer/Data Scientist   
1045  Machine Learning Engineer, Content Intelligenc...   

                                         Position_clean  
286                               Data Scientist Intern  
450                               Senior Data Scientist  
456                                      Data Scientist  
547                                      Data Scientist  
548                                      Data Scientist  
549                                      Data Scientist  
697   Spécialiste, System Design, Data Scientist Mac...  
718                                      Data Scientist  
919                   Data Scientist – Machine Learning  
1025                                     Data Scientist  
1045                                                NaN

In [23]:
# replace

# Senior ds
bool_sr = df_ca['Position_clean'].str.contains('senior|sr', case = False, na = False) &\
        (df_ca['Position_clean'].str.contains('ML', case = True, na = False) |\
        df_ca['Position_clean'].str.contains('machine learn', case = False, na = False) )
df_ca.loc[bool_sr, 'Position_clean'] = SR

# ds
bool_ds =  df_ca['Position_clean'].str.contains('scien', case = False, na = False) &\
        (df_ca['Position_clean'].str.contains('ML', case = True, na = False) |\
        df_ca['Position_clean'].str.contains('machine learn', case = False, na = False) )
df_ca.loc[bool_ds, 'Position_clean'] = DS

In [24]:
## Step 5 : other special cases
df_ca.loc[df_ca['Position'] == 'Programmer/Developer Lead - Sr Data Analyst - Visualization', 'Position_clean'] = SR
df_ca.loc[df_ca['Position'] == 'Senior Statistical Programmer/Statistical Scientist', 'Position_clean'] = SR
df_ca.loc[df_ca['Position'] == 'Junior Statistician/Database Specialist (Research Institute)', 'Position_clean'] = JR
df_ca.loc[df_ca['Position'] == 'Data Analyst (SQL/Python/AI)', 'Position_clean'] = DS

In [25]:
## Step 6
bool_notDS = df_ca['Position_clean'].str.contains('Data Scientist', case = False) == False
#df_ca[bool_notDS]['Position_clean'].iloc[340:360]
df_ca.loc[bool_notDS, 'Position_clean'] = rm

In [26]:
df_ca[bool_notDS]['Position'].unique()

array(['Data Analyst',
       'Automotive Software Data Analyst / Tester - Remote',
       'Junior Data Analyst',
       'Optimization Operational Research Scientist', 'Junior Scientist',
       'Senior Database Developer', 'Master Data Analyst / SAP (Junior)',
       'Data Analyst - Junior', 'Data Analyst (Remote)',
       'Remote Cloud Database developer (Python)',
       'Reporting Consultant & Data Analyst / Cognos Expert',
       'Flight Data Analyst', 'Junior Software Developer',
       'Product Data Analyst', 'Data Analyst II',
       'Entry Level Data Analyst (Bilingual French & English)',
       'Jr. Data Analyst', 'Digital Data Analyst',
       'Data Analyst, Security - 297742', 'Tech - Data Analyst',
       'Data Analyst, Product',
       'Data Analyst Intern, Global Investments IT',
       'Data Analyst - Level 1 x2 (Contract)',
       'Student Intern, Customer Data Analyst',
       'Data Analyst and Reporting Specialist', 'Game Data Analyst',
       'Master Data Analyst /V

In [27]:
# check 
a = df_ca[df_ca['Position'].str.contains('Data Scientist', case = False) == False]
a[['Position', 'Position_clean']]

Position       Position_clean
1                                          Data Analyst                  NaN
2                                          Data Analyst                  NaN
4                          Senior Manager, Data Science  Lead Data Scientist
5     Automotive Software Data Analyst / Tester - Re...                  NaN
6                                   Junior Data Analyst                  NaN
7                                   Junior Data Analyst                  NaN
9           Optimization Operational Research Scientist                  NaN
11                                         Data Analyst                  NaN
12                                     Junior Scientist                  NaN
13                                         Data Analyst                  NaN
14                                         Data Analyst                  NaN
15                                         Data Analyst                  NaN
16                                         Data Analyst                  NaN
17                                         Data Analyst                  NaN
18                                         Data Analyst                  NaN
19                                         Data Analyst                  NaN
20                                         Data Analyst                  NaN
21                                         Data Analyst                  NaN
22                            Senior Database Developer                  NaN
23                                         Data Analyst                  NaN
24                                         Data Analyst                  NaN
25                                         Data Analyst                  NaN
26                                         Data Analyst                  NaN
27                                         Data Analyst                  NaN
28                                         Data Analyst                  NaN
29                   Master Data Analyst / SAP (Junior)                  NaN
30                                         Data Analyst                  NaN
31                                         Data Analyst                  NaN
32                                         Data Analyst                  NaN
33                                Data Analyst - Junior                  NaN
...                                                 ...                  ...
1228                               Product Data Analyst                  NaN
1229                               Product Data Analyst                  NaN
1230                     Data Analyst - Trauma Services                  NaN
1231                             Marketing Data Analyst                  NaN
1232                         Environmental Data Analyst                  NaN
1233                     Digital Marketing Data Analyst                  NaN
1234                                    HR Data Analyst                  NaN
1235                 Data Analyst - Business Operations                  NaN
1236                                    BI/Data Analyst                  NaN
1237                     SAP System Master Data Analyst                  NaN
1238                    Advanced Marketing Data Analyst                  NaN
1239                        Bioinformatics Data Analyst                  NaN
1240                              Business Data Analyst                  NaN
1241                                       Data Analyst                  NaN
1242              Student Intern, Customer Data Analyst                  NaN
1243                             Technical Data Analyst                  NaN
1244                             Marketing Data Analyst                  NaN
1245                                   SAP Data Analyst                  NaN
1246                                Equity Data Analyst                  NaN
1247                     Digital Marketing Data Analyst                  NaN
1248                             Marketing 

In [28]:
# check no. of unique data scientist listings left after removing irrelevant rows
df_ca[df_ca['Position_clean'].str.contains('Data Scientist', case = False) == True]['Position'].unique().shape

(121,)

In [29]:
df_ca['Position'].unique().shape

(615,)

## Out of 615 *unique* Canada listings, get 121 *unique* data scientist (includes: 6 intern, 1 Jr.,  33 Sr.,  19 Lead,  62 data scientist) listings

In [30]:
bool_lds = df_ca['Position_clean'].str.contains('lead|principal|manage|direct|VP|head', case = False, na = False)
#df_ca[bool_lds]['Position'].unique().shape 
df_ca.loc[bool_lds, 'Position_clean'] = LDS

In [31]:
bool_sr = df_ca['Position_clean'].str.contains('senior|sr', case = False, na = False)
#df_ca[bool_sr]['Position'].unique().shape 
df_ca.loc[bool_sr, 'Position_clean'] = SR

In [32]:
bool_jr = df_ca['Position_clean'].str.contains('junior|jr|early|new grad|entry', case = False, na = False)
df_ca[bool_jr]['Position'].unique().shape 
df_ca.loc[bool_jr, 'Position_clean'] = JR

In [51]:
bool_remain = (df_ca['Position_clean'].str.contains('senior', case = False, na = False) == False) &\
                (df_ca['Position_clean'].str.contains('junior', case = False, na = False) == False) &\
                (df_ca['Position_clean'].str.contains('lead', case = False, na = False) == False) &\
                (df_ca['Position_clean'].str.contains('intern', case = False, na = False) == False) &\
                df_ca['Position_clean'].notna()
df_ca.loc[bool_remain, 'Position_clean'] = DS

In [54]:
# check 
df_ca['Position_clean'].unique()
#df_ca[['Position', 'Position_clean']]

array(['Data Scientist', nan, 'Lead Data Scientist',
       'Data Scientist Intern', 'Senior Data Scientist',
       'Junior Data Scientist'], dtype=object)

In [55]:
# write to csv
df_ca.to_csv('elisa_position_ca.csv')

In [56]:
df = pd.read_csv('elisa_position_ca.csv', index_col = 0)
df.shape

(1258, 13)

In [59]:
df.tail(10)

Company  \
1248  Vigorate Digital Solutions   
1249         Cascadia Scientific   
1250             Tandempark Inc.   
1251    Loblaw Companies Limited   
1252                        Moka   
1253           Napoleon Products   
1254                Hudson Rouge   
1255                        Moka   
1256                    Stingray   
1257                 7D Surgical   

                                            Description  \
1248  Who We Are\r\r\nVigorate Digital is made up of...   
1249  Cascadia Scientific provides a SaaS platform i...   
1250  TandemPark Inc. is a maturing startup company ...   
1251  Location:\r\r\n1 Presidents Choice Circle, Bra...   
1252  The Role\r\r\nYou are passionate about consume...   
1253  Job Details\r\r\nDescription\r\r\nTitle: Busin...   
1254  WHO IS GTB?\r\r\n\r\r\nA global, industry-lead...   
1255  The Role\r\r\n\r\r\nYou are passionate about c...   
1256  Are you ready to join a team of marketing and ...   
1257  At 7D Surgical you will find a vibrant team of...   

      Glassdoor Salary Estimate                      Industry  Location  \
1248                        NaN                           NaN   Toronto   
1249                        NaN                           NaN    Surrey   
1250                        NaN                           NaN  Innisfil   
1251                        NaN               Vehicle Dealers  Brampton   
1252                        NaN             Social Assistance  Montreal   
1253                        NaN                           NaN    Barrie   
1254                        NaN       Advertising & Marketing   Toronto   
1255                        NaN  Computer Hardware & Software  Montreal   
1256                        NaN                           NaN  Montreal   
1257                        NaN                           NaN   Toronto   

                    Position  Rating                     Revenue  \
1248  Marketing Data Analyst     NaN                         NaN   
1249        Jr. Data Analyst     NaN                         NaN   
1250        Jr. Data Analyst     NaN                         NaN   
1251  Marketing Data Analyst     3.6          $10+ billion (USD)   
1252  Marketing Data Analyst     3.2    Unknown / Non-Applicable   
1253         BI Data Analyst     3.0  Less than $1 million (USD)   
1254  Marketing Data Analyst     3.6  Less than $1 million (USD)   
1255  Marketing Data Analyst     4.6    Unknown / Non-Applicable   
1256  Marketing Data Analyst     NaN  Less than $1 million (USD)   
1257     Data Analyst Intern     NaN  Less than $1 million (USD)   

                      Sector                    Size                    Type  \
1248                     NaN                     NaN                     NaN   
1249                     NaN                     NaN                     NaN   
1250                     NaN                     NaN                     NaN   
1251                  Retail        10000+ Employees        Company - Public   
1252              Non-Profit    201 to 500 Employees  Nonprofit Organization   
1253                     NaN  1001 to 5000 Employees       Company - Private   
1254       Business Services     51 to 200 Employees       Company - Private   
1255  Information Technology    201 to 500 Employees       Company - Private   
1256                     NaN       1 to 50 Employees       Company - Private   
1257                     NaN       1 to 50 Employees       Company - Private   

      Year Founded Position_clean  
1248           NaN            NaN  
1249           NaN            NaN  
1250           NaN            NaN  
1251           NaN            NaN  
1252        1978.0            NaN  
1253           NaN            NaN  
1254           NaN            NaN  
1255        2014.0            NaN  
1256           NaN            NaN  
1257           NaN            NaN